overlord/patch/patch5.go

```diff
@@ -0,0 +1,81 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patch
+
+import (
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/wrappers"
+)
+
+func init() {
+	patches[5] = patch5
+}
+
+type log struct{}
+
+func (log) Notify(status string) {
+	logger.Noticef("patch 5: %s", status)
+}
+
+func patch5(st *state.State) error {
+	log := log{}
+
+	snapStates, err := snapstate.All(st)
```

Comment:
We have a bug in the way we're producing our patches. Imagine next week we add patch 6, which changes the format of SnapState. Suddenly patch 5 breaks down because it's doing an earlier migration.

At some point we'll either need to create local minimal patch infrastructure which reflects the intermediate states being manipulated at the time, or we'll need to break compatibility permanently with older releases, and force people to go through an older snapd (hard for apt).